Running Quantized ResNet50 via MIGraphX

Rmalavally opened the pull request:
https://ontrack-internal.amd.com/browse/SWDEV-427204

https://ontrack-internal.amd.com/browse/ROCDOC-210

Rmalavally committed:
Create HIP_example

Rmalavally committed:
Merge pull request #1 from Rmalavally/Rmalavally-patch-1

Create HIP_example

Rmalavally committed:
Create Using-Address-Sanitizer-with-a-Short-HIP-Application.md

Rmalavally committed:
Update Using-Address-Sanitizer-with-a-Short-HIP-Application.md

Rmalavally committed:
Delete HIP_example

Rmalavally committed:
Merge branch 'amd:develop' into develop

Rmalavally committed:
Create Quantizations.rst

Adding new directory for Frameworks integrations/Quantization example for INT8 update

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Updated to .rst

Rmalavally committed:
Update Quantizations.rst

Added note

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Added prerequisite

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

Rmalavally committed:
Update Quantizations.rst

emankov reviewed (approved):
LGTM

saadrahim reviewed (changes requested):
There are two changes included in this PR. They should be separated. I am not sure what the goal is of this yet. Frameworks-integrations should not be the directory name. Directory name needs some thought.

Rmalavally committed:
Delete LLVM_ASAN/Using-Address-Sanitizer-with-a-Short-HIP-Application.md

Rmalavally commented:
> There are two changes included in this PR. They should be separated. I am not sure what the goal is of this yet. Frameworks-integrations should not be the directory name. Directory name needs some thought.

Thanks, Saad. Deleted the ASAN file. 

'Frameworks integrations' will include examples of ROCm applications in PyTorch, Tensorflow, and others. The recommendation came from the MIGraphX team. Please let me know if you have any suggestions. @saadrahim @shivadbhavsar

LisaDelaney commented:
@Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html

saadrahim commented on Frameworks-integrations/MIGraphX/Quantizations.rst:
```diff
@@ -0,0 +1,145 @@
+Running Quantized ResNet50 via MIGraphX
+========================================
+
+This example walks through the dynamo PTQ workflow for running a quantized model using torch_migraphx.
```
What is a PTQ?

Rmalavally replied:
Post Training Quantization. Corrected. @saadrahim

saadrahim commented on Frameworks-integrations/MIGraphX/Quantizations.rst:
```diff
@@ -0,0 +1,145 @@
+
+    print(f"{mgx_fp32_time=:0.4f}ms")
+    print(f"{mgx_fp16_time=:0.4f}ms")
+    print(f"{mgx_int8_time=:0.4f}ms")
```
Please conclude this example

saadrahim replied:
What was the performance gain? I don't want to run to find out, I'd like to know when I am reading.

Rmalavally replied:
@LisaDelaney @saadrahim Let's discuss the format for examples. Adding much content will change this from an example to a blog/article. I am ok if we want to convert this to a blog. 

@emankov @shivadbhavsar

Rmalavally replied:
> What was the performance gain? I don't want to run to find out, I'd like to know when I am reading.

@shivadbhavsar Please advise.

saadrahim commented on Frameworks-integrations/MIGraphX/Quantizations.rst:
```diff
@@ -0,0 +1,145 @@
+Prerequisite 
++++++++++++++++
+
+You must follow the `installation instructions <https://github.com/ROCmSoftwarePlatform/torch_migraphx>`_ before using this example. 
```
installation instructions for what?

Rmalavally replied:
I have replicated the installation instructions in the Readme file and assume any changes to the original instructions will also be reflected here. Otherwise, it could be a maintenance task. Please advise.  @saadrahim @emankov @shivadbhavsar

shivadbhavsar replied:
Installation instructions for the `torch_migraphx` library (where the link points)

Rmalavally replied:
> installation instructions for what?

Updated.

saadrahim commented on Frameworks-integrations/MIGraphX/Quantizations.rst:
Folder can be AI?

Rmalavally replied:
Not sure about AI. Let me get back to you on this.

Rmalavally replied:
Sounds good, @saadrahim. Thanks for the suggestion.

saadrahim commented on Frameworks-integrations/MIGraphX/Quantizations.rst:
```diff
@@ -0,0 +1,145 @@
+    q_m = convert_pt2e(m)
+    torch_qout = q_m(input_fp32)
+
+2. Lower Quantized model to MIGraphX
```
The meaning of this title does not make sense to me. import lower quantized model, ingest lower quantized model, build lower quantized model?

Rmalavally replied:
This example is reviewed and approved by the MIGraphX team. 

@shivadbhavsar, Could you please provide context?

shivadbhavsar replied:
"Lower" here is used as a verb. Its a commonly used term in the context of optimizing deeplearning models on specific hardware. It means this step is "lowering" the model from the framework level to the graph compiler level

saadrahim reviewed (changes requested):
This needs to be added in the TOC of the main README.md file in the repo.

It needs to be reviewed by the MIGraphX team.

What is the purpose of this example? In general, I am not sure if this is an example or if it should be some form of blog.

saadrahim commented:
> @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html

This repo does not use RST as it has no generated documentation. Please convert to markdown.

Rmalavally commented:
> @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html

Does the style guide apply to ROCm examples as well? Please confirm.

Rmalavally commented:
> > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> 
> This repo does not use RST as it has no generated documentation. Please convert to markdown.

I can change it to Markdown, not a problem. 

@LisaDelaney - The Libraries examples have Summary and Prerequisites, while the HIP examples have Description and Key APIs. Which one should we follow for examples? There is no consistency.

Each folder has examples in different styles. My recommendation is to create a simple format (we could finalize on either the Libraries or HIP formats) for all examples.

Rmalavally commented:
> > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> 
> Does the style guide apply to ROCm examples as well? Please confirm.

Using the contributing guide will make it inconsistent with other examples.

Rmalavally commented:
> > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> 
> This repo does not use RST as it has no generated documentation. Please convert to markdown.

@saadrahim this is the guideline in our style guide. Please advise. 

>Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.

Rmalavally committed:
Delete Frameworks-integrations/MIGraphX/Quantizations.rst

Rmalavally committed:
Merge branch 'amd:develop' into Rmalavally-patch-2-INT8-ROCm6.1

Rmalavally committed:
Create test.md

Rmalavally committed:
Create test.md

Rmalavally committed:
Create test.md

Rmalavally committed:
Add files via upload

Rmalavally committed:
Delete AI/test.md

Rmalavally committed:
Delete AI/MIGraphX/test.md

Rmalavally committed:
Delete AI/MIGraphX/Quantization/test.md

Rmalavally committed:
Delete AI/MIGraphX/Quantization/Readme.md

Rmalavally committed:
Add files via upload

Rmalavally committed:
Delete AI/MIGraphX/Readme.md

Rmalavally committed:
Add files via upload

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

shivadbhavsar commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -0,0 +1,112 @@
+  
+- Dynamo Backend: Importing torch_migraphx automatically registers the "migraphx" backend that can be used with the `torch.compile` API.
+
+
```
Im not sure if it's a good idea to copy paste the readme from torch_migraphx rather than just link to it

Rmalavally replied:
Agreed and corrected.

shivadbhavsar commented on AI/MIGraphX/Quantization/Readme.md:
````diff
@@ -0,0 +1,112 @@
+export TORCH_CMAKE_PATH=$(python -c "import torch; print(torch.utils.cmake_prefix_path)")
+pip install .
+```
+
````
Even if the decision is to leave in the pasted installation instructions, the content below this line is probably not necessary since the examples in this repo themselves will show the user how to use the provided APIs in this context.

Rmalavally replied:
@shivadbhavsar I added a link to the Torch_MIGraphX page.

shivadbhavsar commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
````diff
@@ -0,0 +1,105 @@
+```
+Use the pt2e API to prepare, calibrate, and convert the model. Torch-MIGraphX provides a custom Quantizer for performing quantization that is compatible with MIGraphX.
+
+**Note**: Additional configs will also work as long as the configs ensure symmetric quantization using the signed int8 datatype. Currently, only symmetric quantization is supported in MIGraphX.
````
Remove this note, its not relevant anymore.

Rmalavally replied:
Removed.

Rmalavally replied:
@saadrahim The comments have been resolved. Please let me know if you have any other recommendations. If not, could you please unblock for merging? Thanks so much.

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

LisaDelaney commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
```diff
@@ -60,7 +60,7 @@ print(f"MIGraphX INT8:\n{mgx_out}")
 
 3. Performance
-Let’s do a quick test to measure the performance gain from using quantization. Note that these performance gains (or lack of gains) will vary depending on the specific hardware in use.
+Let’s do a quick test to measure the performance gain from using quantization.
```
```suggestion
Do a quick test to measure the performance gain from using quantization.
```

LisaDelaney commented:
@Rmalavally there are a couple titles not in sentence case--mind fixing those?

LisaDelaney commented:
> > > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> > 
> > 
> > This repo does not use RST as it has no generated documentation. Please convert to markdown.
> 
> @saadrahim this is the guideline in our style guide. Please advise.
> 
> > Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.

@Rmalavally if this repo is intended to be viewed mostly from GitHub (which I think it is), markdown makes more sense in this case. If this changes in the future, we can sort it out then.

Rmalavally commented:
> > > > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> > > 
> > > 
> > > This repo does not use RST as it has no generated documentation. Please convert to markdown.
> > 
> > 
> > @saadrahim this is the guideline in our style guide. Please advise.
> > > Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.
> 
> @Rmalavally if this repo is intended to be viewed mostly from GitHub (which I think it is), markdown makes more sense in this case. If this changes in the future, we can sort it out then.

Sounds good. Please add this guideline to the style guide, or I could too.

Rmalavally commented:
> > > > > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> > > > 
> > > > 
> > > > This repo does not use RST as it has no generated documentation. Please convert to markdown.

Please update the style guide with this detail.
> > > 
> > > 
> > > @saadrahim this is the guideline in our style guide. Please advise.
> > > > Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.
> > 
> > 
> > @Rmalavally if this repo is intended to be viewed mostly from GitHub (which I think it is), markdown makes more sense in this case. If this changes in the future, we can sort it out then.
> 
> Sounds good. Please add this guideline to the style guide, or I could too.

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Minor edits

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

LisaDelaney commented:
> > > > > > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> > > > > 
> > > > > 
> > > > > This repo does not use RST as it has no generated documentation. Please convert to markdown.
> 
> Please update the style guide with this detail.
> 
> > > > @saadrahim this is the guideline in our style guide. Please advise.
> > > > > Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.
> > > 
> > > 
> > > @Rmalavally if this repo is intended to be viewed mostly from GitHub (which I think it is), markdown makes more sense in this case. If this changes in the future, we can sort it out then.
> > 
> > 
> > Sounds good. Please add this guideline to the style guide, or I could too.

This does not need to be added to our style guide. You can make decisions based on context using critical thinking and logical thought. As a tech writer, minor/once-off exceptions are at your discretion; these do not need to clutter up a style guide.

Rmalavally commented:
> > > > > > > @Rmalavally the syntax isn't consistent with our style guide. Can you please update it? See: https://rocm.docs.amd.com/en/latest/contribute/contribute-docs.html
> > > > > > 
> > > > > > 
> > > > > > This repo does not use RST as it has no generated documentation. Please convert to markdown.
> > 
> > 
> > Please update the style guide with this detail.
> > > > > @saadrahim this is the guideline in our style guide. Please advise.
> > > > > > Our documentation includes both Markdown and RST files. We are gradually transitioning existing Markdown to RST in order to more effectively meet our documentation needs. When contributing, RST is preferred; if you must use Markdown, use GitHub-flavored Markdown.
> > > > 
> > > > 
> > > > @Rmalavally if this repo is intended to be viewed mostly from GitHub (which I think it is), markdown makes more sense in this case. If this changes in the future, we can sort it out then.
> > > 
> > > 
> > > Sounds good. Please add this guideline to the style guide, or I could too.
> 
> This does not need to be added to our style guide. You can make decisions based on context using critical thinking and logical thought. As a tech writer, minor/once-off exceptions are at your discretion; these do not need to clutter up a style guide.

In general, i think good for consistency as it digresses from our standard practices. I am good with all options.

saadrahim commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -0,0 +1,50 @@
+
+## Introduction
+
+Quantization is an optimization tool that is commonly used to accelerate inferencing for deep learning models. There are many flavors of quantization. In the examples here, we show how the ROCm stack can be used to accelerate inference using [INT8 quantization](https://arxiv.org/pdf/1712.05877.pdf) on pretrained models. 
```
```suggestion
Quantization is an optimization tool that is commonly used to accelerate inferencing for deep learning models. There are many flavors of quantization. In the examples here, we show how the ROCm can be used to accelerate inference using [INT8 quantization](https://arxiv.org/pdf/1712.05877.pdf) on pretrained models. 
```

saadrahim commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
```diff
@@ -0,0 +1,104 @@
+This example walks through the dynamo Post Training Quantization (PTQ) workflow for running a quantized model using torch_migraphx.
+
+## Prerequisites
+You must follow the installation instructions for the torch_migraphx library in [Readme.md](https://github.com/Rmalavally/rocm-examples/blob/develop/AI/MIGraphX/Quantization/Readme.md) before using this example.
```
personal forks should not be linked to.

Rmalavally replied:
Will update this once we have the approval to merge.

Rmalavally replied:
Added. Please review the Readme file in rocm-examples.

saadrahim commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -0,0 +1,50 @@
+# MIGraphX - Torch Examples
+
+# Summary
+The examples in this subdirectory showcase the functionality of the [MIGraphX](https://github.com/ROCmSoftwarePlatform/torch_migraphx/tree/master) library. The examples build on Linux for the ROCm (AMD GPU) backend.
```
```suggestion
The examples in this subdirectory showcase the functionality of the [MIGraphX](https://github.com/ROCmSoftwarePlatform/torch_migraphx/tree/master) library. The examples build on Linux for the ROCm (AMD GPU) backend.
```
 Last sentence is too technical, let's rewrite.  AMD GPU backend term does not need to be used.

Rmalavally replied:
Fixed. Please review the updated Summary section in the Readme file.

Rmalavally committed:
Update Readme.md

removed 'stack'

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

shivadbhavsar commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -0,0 +1,50 @@
+# MIGraphX - Torch Examples
+
+# Summary
+The examples in this subdirectory showcase the functionality of the [MIGraphX](https://github.com/ROCmSoftwarePlatform/torch_migraphx/tree/master) library. The examples build on Linux for the ROCm (AMD GPU) backend.
```
For the link please change text to Torch-MIGraphX from just MIGraphX. Thats confusing because MIGraphX is its own repo (not the one we are pointing to in this case)

Rmalavally committed:
Update Readme.md

Rmalavally committed:
Update Readme.md

Updated the link to Torch-MIGraphX

Rmalavally committed:
Update Readme.md

Summary update

Rmalavally committed:
Update README.md

Updated Readme.md with the relative link

LisaDelaney commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -6,7 +6,7 @@ The examples in this subdirectory showcase the functionality of the [MIGraphX](h
 ## Introduction
 
-Quantization is an optimization tool that is commonly used to accelerate inferencing for deep learning models. There are many flavors of quantization. In the examples here, we show how the ROCm stack can be used to accelerate inference using [INT8 quantization](https://arxiv.org/pdf/1712.05877.pdf) on pretrained models. 
+Quantization is an optimization tool that is commonly used to accelerate inferencing for deep learning models. There are many flavors of quantization. In the examples here, we show how the ROCm can be used to accelerate inference using [INT8 quantization](https://arxiv.org/pdf/1712.05877.pdf) on pretrained models. 
```
```suggestion
Quantization is an optimization tool that is commonly used to accelerate inferencing for deep learning models. There are many flavors of quantization. In the examples here, we show you how to use ROCm to accelerate inference using [INT8 quantization](https://arxiv.org/pdf/1712.05877.pdf) on pretrained models. 
```

Rmalavally committed:
Update README.md

shivadbhavsar commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
```diff
@@ -0,0 +1,105 @@
+
+The capture_pre_autograd_graph call will be changed to a torch.export.export call once it supports the pre autograd capture functionallity. 
+
+**Note**: Currently, there is a known issue when using only kwargs as inputs. See pytorch/pytorch#113744 for more information.
```
This line can be removed, its not relevant to the model in this example

shivadbhavsar commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
````diff
@@ -0,0 +1,105 @@
+torch_fp32_out = model_fp32(input_fp32)
+```
+
+The capture_pre_autograd_graph call will be changed to a torch.export.export call once it supports the pre autograd capture functionallity. 
````
This can be removed, it will probably just confuse the user. We will need to update this example when/if PyTorch makes this change.

Rmalavally replied:
Fixed

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Removed the following note,

**Note**: Currently, there is a known issue when using only kwargs as inputs. See pytorch/pytorch#113744 for more information.
@shivadbhavsar shivadbhavsar 12 minutes ago
This line can be removed, its not relevant to the model in this example

Rmalavally committed:
Update Running Quantized ResNet50 via MIGraphX.md

Removed the following line,

The capture_pre_autograd_graph call will be changed to a torch.export.export call once it supports the pre autograd capture functionallity.

causten commented on AI/MIGraphX/Quantization/Readme.md:
```diff
@@ -0,0 +1,50 @@
+
+### Installation instructions
+
+Refer to the [Torch_MIGraphX](https://github.com/ROCmSoftwarePlatform/torch_migraphx/blob/master/README.md) page for Docker and source installation instructions. 
```
The repo moved to https://github.com/ROCm/torch_migraphx/blob/master/README.md

causten commented on AI/MIGraphX/Quantization/Running Quantized ResNet50 via MIGraphX.md:
```diff
@@ -0,0 +1,102 @@
+
+## Prerequisites
+
+- You must follow the installation instructions for the torch_migraphx library in [Readme.md](https://github.com/Rmalavally/rocm-examples/blob/develop/AI/MIGraphX/Quantization/Readme.md) before using this example.
```
Change to the correct org

Rmalavally replied:
@causten The content in Readme.md accompanying the example is different from the Readme in the repo. 

We have the following link at the end of the file for installation instructions,

>Refer to the [Torch_MIGraphX](https://github.com/ROCmSoftwarePlatform/torch_migraphx/blob/master/README.md) page for Docker and source installation instructions.

@shivadbhavsar Please confirm.

Rmalavally replied:
@causten Let me know if you have any concerns. The Readme with the example provides contextual details about the example. If you are ok with this, can we merge the PR? Please advise.
